core/arch/x86/x86.asm

```diff
@@ -2284,10 +2284,96 @@ GLOBAL_LABEL(get_ymm_caller_saved:)
         ret
         END_FUNC(get_ymm_caller_saved)
 
-/* void hashlookup_null_handler(void)
- * PR 305731: if the app targets NULL, it ends up here, which indirects
- * through hashlookup_null_target to end up in an ibl miss routine.
+/* void get_zmm_caller_saved(byte *zmm_caller_saved_buf)
+ *   stores the values of zmm0 through zmm31 consecutively into zmm_caller_saved_buf.
+ *   zmm_caller_saved_buf need not be 64-byte aligned.
+ *   caller must ensure that the underlying processor supports AVX-512!
  */
+        DECLARE_FUNC(get_zmm_caller_saved)
+GLOBAL_LABEL(get_zmm_caller_saved:)
+        mov      REG_XAX, ARG1
+       /* i#441: some compilers like gcc 4.3 and VS2005 do not know "vmovdqu"
+        * or need architectural flags set (e.g. -mavx or -march=skylake-avx512),
+        * which would cause DynamoRIO to be less (or un-) portable.
+        * We just put in the raw bytes for these instrs:
+        * Note the 64/32 bit have the same encoding for either rax or eax.
+        * Trivial: the destination (memory-) operand implies mask register k0.
```

Comment:
If this depends on k0, are there ABI conventions about k0 that make that ok?  Can we rely on it having a certain value, or does the caller have to set it?

Reply:
Most (or all? ... not sure)  AVX-512 instructions have an additional operand, vector mask register applied to the destination 'write'. They fixed the k0 mask register to all ones == unmasked in hardware. Maybe in order to simplify, most (or all? ... not sure) assemblers don't allow k0 in the mnemonic. If you mean k0, you don't write it. I was just pointing this out, but I guess this is irrelevant here, I've removed the comment...!

Reply:
See previous comment.